File: Python/SwapNodesInPairs.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        nxt = head.next
        head.next = self.swapPairs(nxt.next)
        nxt.next = head
        return nxt


def build(vals):
    dummy = ListNode()
    cur = dummy
    for v in vals:
        cur.next = ListNode(v)
        cur = cur.next
    return dummy.next


def to_list(head):
    out = []
    while head:
        out.append(head.val)
        head = head.next
    return out


if __name__ == '__main__':
    s = Solution()
    print(to_list(s.swapPairs(build([1, 2, 3, 4]))))  # [2, 1, 4, 3]
    print(to_list(s.swapPairs(build([]))))            # []
    print(to_list(s.swapPairs(build([1]))))           # [1]
    print(to_list(s.swapPairs(build([1, 2, 3]))))     # [2, 1, 3]

File: Java/SwapNodesInPairs.java
import java.util.ArrayList;
import java.util.List;

public class SwapNodesInPairs {
    static class ListNode {
        int val;
        ListNode next;
        ListNode(int v) { this.val = v; }
    }

    public ListNode swapPairs(ListNode head) {
        if (head == null || head.next == null) return head;
        ListNode nxt = head.next;
        head.next = swapPairs(nxt.next);
        nxt.next = head;
        return nxt;
    }

    static ListNode build(int... vals) {
        ListNode dummy = new ListNode(0);
        ListNode cur = dummy;
        for (int v : vals) {
            cur.next = new ListNode(v);
            cur = cur.next;
        }
        return dummy.next;
    }

    static List<Integer> toList(ListNode h) {
        List<Integer> out = new ArrayList<>();
        while (h != null) { out.add(h.val); h = h.next; }
        return out;
    }

    public static void main(String[] args) {
        SwapNodesInPairs s = new SwapNodesInPairs();
        System.out.println(toList(s.swapPairs(build(1, 2, 3, 4)))); // [2, 1, 4, 3]
        System.out.println(toList(s.swapPairs(build())));           // []
        System.out.println(toList(s.swapPairs(build(1))));          // [1]
        System.out.println(toList(s.swapPairs(build(1, 2, 3))));    // [2, 1, 3]
    }
}
